src/backend/commands/sequence.c

```diff
@@ -966,6 +969,23 @@ do_setval(Oid relid, int64 next, bool iscalled)
 	/* lock page' buffer and read tuple */
 	seq = read_seq_tuple(seqrel, &buf, &seqdatatuple);
 
+	/*
+	 * BABELFISH: In T-SQL, with identity_insert=on, sequence value is set as
+	 * max(value to be inserted, last used sequence value) when increment is
+	 * positive. Min value is set with a negative increment.
+	 * We need to calculate the max/min and set the value without releasing
+	 * the lock so that other backends can't overwrite the value concurrently.
+	 */
+	if (sql_dialect == SQL_DIALECT_TSQL)
```

Comment:
First of all, could you introduce a hook to minimize the engine changes? At least, we don't need to add `parser.h` and `seqincrement`.

Secondly, we should think more carefully on the dialect check.
1. What would be our expectation when a user tries to insert in PG connection against a T-SQL table? You may need to sync up with Di regarding those kind of inter-ops behavior.
2. In the future, we would want to support data dump and restore. Suppose we have two tables in the BBF database, t1 created from PG connection and t2 created from TDS connection. In that case, how can we distinguish those two tables?

Reply:
> First of all, could you introduce a hook to minimize the engine changes? At least, we don't need to add parser.h and seqincrement.
You're right. I should add a function hook instead.

> Secondly, we should think more carefully on the dialect check.
> 
> 1. What would be our expectation when a user tries to insert in PG connection against a T-SQL table? You may need to sync up with Di regarding those kind of inter-ops behavior.
These changes only affect whether the sequence seed update should be transactional or not in TSQL dialect. As per the current implementation,
1. From TDS connection, if you insert an identity column, the seed update is transactional. That means if the insert rollbacks, the seed will not be updated with the new value. This is not the SQL Server behaviour.
2. Internally, this is implemented as 'GENERATED ALWAYS AS IDENTITY'. From PG connection, you're not allowed to insert directly in an identity column. It'll throw an error. There are ways to insert from PG connection as well - a) using OVERRIDING SYSTEM VALUE with INSERT b) using COPY. But, both these option doesn't update the seed by default. User has to manually update the seed using setval function. In fact, pg_dump uses the same approach for dumping identity columns. An example:

```
-- Table created from TDS connection
CREATE TABLE t1(a int IDENTITY, b int);
INSERT INTO t1 VALUES(1);

-- the pg_dump script for the table (pg_dump jdbc_testdb --table=master_dbo.t1 --column-inserts)
CREATE TABLE master_dbo.t1 (
    a integer NOT NULL,
    b integer
);
ALTER TABLE master_dbo.t1 ALTER COLUMN a ADD GENERATED ALWAYS AS IDENTITY (
    SEQUENCE NAME master_dbo.t1_a_seq
    START WITH 1
    INCREMENT BY 1
    NO MINVALUE
    NO MAXVALUE
    CACHE 1
);
INSERT INTO master_dbo.t1 (a, b) OVERRIDING SYSTEM VALUE VALUES (1, 1);
SELECT pg_catalog.setval('master_dbo.t1_a_seq', 1, false);
```

pg_dump calculates the seed as following:
```
select max(a) from master_dbo.t1
```
and generates the setval statement in the dump.

In this patch, there will be no behavioural changes in PG connection. Only from TDS connection, the seed update will become non-transactional similar to T-SQL and PG.

> 2. In the future, we would want to support data dump and restore. Suppose we have two tables in the BBF database, t1 created from PG connection and t2 created from TDS connection. In that case, how can we distinguish those two tables?

As I explained above, with these changes, the behaviour will be now same for TDS and PG connections. So, I don't think we need to distinguish the connection type at least from this feature's perspective.